Selectable Header in Component Group View.

farmaazon opened the pull request:
[ci no changelog needed]

### Pull Request Description

* The List View component was refactored: it allows for hiding the internal selection widget, and exposes information where the widget should be placed. This allows us to create selection widget in component list panel, so it can be animated between component groups and sections.
* Fixed some warnings when checking WASM code.
* Adjusted the style of Component Group View a little, so it better reflects the design doc. Still not ideal, because the list_view has some weird design regarding padding, but I don't want to stuck in some bigger refactoring.

I will add a video in a few minutes.

### Important Notes

https://user-images.githubusercontent.com/3919101/165507826-60329f9e-7de3-4eb2-9271-292e45568cb2.mov


### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

farmaazon committed:
Options for ListView

farmaazon committed:
WIP

farmaazon committed:
WIP

farmaazon committed:
Fixes

farmaazon committed:
Self-review

farmaazon committed:
Merge remote-tracking branch 'origin/develop' into wip/farmaazon/cgv-selectable-header-181725016

farmaazon committed:
Linter

wdanilo commented on app/gui/view/component-browser/component-group/src/lib.rs:
```diff
@@ -17,12 +17,15 @@
 use enso_frp as frp;
+use ensogl_core::application::shortcut::Shortcut;
 use ensogl_core::application::Application;
+use ensogl_core::application::View as TRAIT_View;
```
this looks ugly. What about `use ensogl_core::application::traits::*` instead?

wdanilo commented on app/gui/view/component-browser/component-group/src/lib.rs:
```diff
@@ -40,6 +43,13 @@ const HEADER_FONT: &str = "DejaVuSans-Bold";
+// === Aliases ===
+// ===============
+
+type EntryId = list_view::entry::Id;
```
why not importing `entry` and using `entry::Id` instead?

farmaazon replied:
Not everything from this entry module is related to component group. Perhaps in the future, when we will have more component-group- entry-specific stuff, we will have our own `entry` module there and `Id` reimported there.

wdanilo replied:
Add there short explanation about it, pls!

wdanilo commented on app/gui/view/component-browser/component-group/src/lib.rs:
```diff
@@ -65,6 +75,19 @@ pub mod background {
+    ensogl_core::define_shape_system! {
+        above = [background];
+        () {
+            let bg_color = Rgba::new(0.0,0.0,0.0,0.000_001);
```
There was a constant for that – please use it instead, as this looks like a magic :D

farmaazon replied:
I cannot find it. There is only `transparent` method of `Rgba`, but there the alpha is 0.0. Shall I create such constant (or `Rgba`'s static method)?

wdanilo replied:
<img width="625" alt="image" src="https://user-images.githubusercontent.com/1623053/165511728-8cd4a67d-7f41-442f-8467-36512729e463.png">

farmaazon replied:
Pfff, why _red_? I was looking for black :P

wdanilo commented on app/gui/view/component-browser/component-group/src/lib.rs:
```diff
@@ -244,6 +345,9 @@ impl Model {
 ///
 /// The widget is rendered as a header label, a list of entries below it, and a colored background.
+/// It does not display the selection widget - because the selection jump between various component
+/// groups, the parent (Component List Panel) should own the selection widget; the Component Group
+/// View provide the information, where the widget should be placed when it's focused.
```
- provide -> provides
- remove `,` before where
- Sometimes "Component Group" is written with upper case, sometimes with lower case

farmaazon replied:
Switched to upper case. BTW, the design doc is inconsistent too, however the upper case wins with over-10 to 3 score.

wdanilo commented on lib/rust/ensogl/component/list-view/src/lib.rs:
```diff
@@ -128,6 +128,21 @@ struct View {
     size:       Vector2<f32>,
 }
 
+/// An internal structure describing where selection would go after jump, i.e. after navigating with
+/// arrows or PgUp/PgDown.
+#[derive(Copy, Clone, Debug)]
```
This doc des not explain what "Above" nor "Below" is. Is it above / below the current selection?

wdanilo commented:
\<waiting for video\>

farmaazon committed:
Apply @wdanilo review (and forgotten limit bump)

farmaazon committed:
Merge branch 'develop' into wip/farmaazon/cgv-selectable-header-181725016

farmaazon committed:
Additional review round

farmaazon commented:
@akavel ready for testing.

farmaazon committed:
Merge branch 'develop' into wip/farmaazon/cgv-selectable-header-181725016

akavel commented on app/gui/view/component-browser/component-group/src/lib.rs:
```diff
@@ -141,7 +185,7 @@ impl component::Frp<Model> for Frp {
             );
             eval input.set_background_color((c)
-                model.background.color.set(c.into()));
+            model.background.color.set(c.into()));
```
Why do both of the lines above have the same indentation? Shouldn't the 2nd line be indented compared to the line above it, given that both of the lines are parts of the same `eval` statement?

akavel reviewed (changes requested):
I noticed the following issues while doing the Acceptance/QA Review of the task. Based on that, I'll currently set the Review as "Revise".

### 1
When the `component_group` demo scene is first opened, the selection shape looks incorrectly aligned to the selected entry's text (too high) - see the image below:

<img width="202" alt="Screenshot 2022-04-28 at 11 27 18" src="https://user-images.githubusercontent.com/273837/165740240-1778b88b-8b68-4772-9c66-f5dab4d9ec10.png">

After hovering the same item with mouse, the shape looks correctly aligned:

<img width="188" alt="Screenshot 2022-04-28 at 11 27 44" src="https://user-images.githubusercontent.com/273837/165740289-a6cd9755-2124-4e05-b599-7421b00e9463.png">




### 2

When the `component_group` demo scene is first opened, pressing the `down` key on the keyboard results in the first entry ("long sample ent...") becoming focused. Given that "table input" looks highlighted initially, the expected behavior after pressing the `down` key would be that the "text input" entry becomes selected instead.

https://user-images.githubusercontent.com/273837/165740608-cbe3121b-e305-4fd6-a345-be5c08332e22.mov



### 3

In Node Searcher, the selection does not appear highlighted. This seems to be a regression from commit 88ec8c477.

https://user-images.githubusercontent.com/273837/165740786-8243eb24-56d6-474a-a96e-32cb22e76ee2.mov

### 4

AC Scenario 4 does not appear to be implemented, and I don't see an explanation for a reason for that. Notably, https://www.pivotaltracker.com/story/show/181725003 seems to be a requirement for implementing AC Scenario 4.

### 5

Requirement of AC Scenario 1 does not appear to be fulfilled - I don't see a message in the debug console when I click the group name with the mouse.

### 6

The Design Doc defines the "Suggestion Acceptance Key" to be the `tab` key by default. Requirement of AC Scenario 1 does not appear to be fulfilled - I don't see a message in the debug console when I press the `tab` key while an item is selected using the up/down keys on the keyboard in the Component Group view.

### 7

When I select an entry using the up/down keys on the keyboard, and then click the mouse hovering over a different entry (without moving the mouse), a message for the mouse-hovered entry is printed in the console, but the item hovered with the mouse does not become selected. This behavior does not seem intuitive. See the video below.


https://user-images.githubusercontent.com/273837/165741136-32095e51-7563-49ad-a001-578cc1373f24.mov



### 8

I noticed a corner case that I did not find clear guidance for in the design doc; it's not clear to me if the current behavior is correct, so I'd like to see it clarified. Specifically, to reproduce:
1. Press the `down` key on the keyboard 8 times (to ensure that the entries are scrolled).
2. Hover the mouse over the top-most partially visible entry (the one below the group name).
3. Observe, that the selection shape partially overlaps the group name background area, as shown in the image below:

<img width="174" alt="Screenshot 2022-04-28 at 12 54 09" src="https://user-images.githubusercontent.com/273837/165740389-16157def-caa9-419d-875c-855f0ae2da25.png">

This becomes more clearly visible when we reduce the height of the component_group widget in the debug scene to 190:
```
    // component_group.set_size(Vector2(150.0, 200.0));
    component_group.set_size(Vector2(150.0, 190.0));
```
as seen in the video below:

https://user-images.githubusercontent.com/273837/165741501-4072ffff-f45a-4da8-bb79-abc7cf54bbc4.mov


It's especially unclear to me how will the Component Group look in this case after https://www.pivotaltracker.com/story/show/181725003 is merged. In my opinion, the preferred behavior here would be that the list of the entries would be scrolled to make the partially visible item fully visible when hovered with the mouse. A similar behavior could be implemented when the bottom item is only partially visible and becomes hovered with the mouse (as seen at the beginning of the video above). However, I can accept the current behavior if it's clarified that it is the expected one, and we ensure that the look and feel in this scenario are sensible after both this story and https://www.pivotaltracker.com/story/show/181725003 are merged.

### 9

After hovering an item using the mouse, then pressing `up` key on the keyboard multiple times (such that the group name becomes selected), then pressing the `down` key on the keyboard once, sometimes the 2nd entry (instead of the 1st one) becomes selected. In the video below, during the first attempt at reproducing, the behavior is correct, but during the second attempt, the behavior is incorrect. I noticed the incorrect behavior more frequently than the correct one when reproducing, I'm not sure what circumstances differentiate them.

https://user-images.githubusercontent.com/273837/165741827-49f7553c-0a4c-4b32-b9a8-62d1e04ba761.mov

farmaazon committed:
Merge branch 'develop' into wip/farmaazon/cgv-selectable-header-181725016

farmaazon committed:
Fixes after QA round

farmaazon committed:
One more fix and missing tests

farmaazon commented:
Ad 1,2,3: Fixed
Ad 4: I missed this one. It does not seem possible to implement it right now. Can we move this AC to that issue?
Ad 5,6,7: Fixed
Ad 8: I "fixed" that by making the group size depending on number of entries; In case when the group is partially scrolled, we should just move the header (the part over the header will be covered). @vitvakatu it touches your task here.
Ad 9: Fixed

akavel commented:
@farmaazon 

Ad 4: I'm OK if you move the AC Scenario 4 elsewhere.

Ad 7: This does not appear to be fixed - see the video below:

https://user-images.githubusercontent.com/273837/165916401-b35a46a7-3702-4a17-b4a1-b182c9cacd92.mov

Ad 8: As I wrote above, I'm OK if we solve the interaction as part of some other ticket. I'd only want to request that it's explicitly stated as part of the description of the ticket (whether in AC or in the Implementation Notes) so that we don't forget about it during QA/Acc Review.

As for some additional things I noticed:

### 10

When pressing `enter` after selecting the group name using the keyboard, I don't see a message in the console (I see it when pressing `tab`). Is this by design or an issue? I must admit I'm not completely sure what's the expected behavior based on the Design Doc.

### 11

Behavior when pressing `up` on the keyboard multiple times in ListView is changed from `develop`. Notably, this changed behavior shows up in the Node Searcher.

Observed behavior with the PR:

https://user-images.githubusercontent.com/273837/165919079-e460272e-f96c-42b1-a7a8-942dfaf95a5c.mov

Observed behavior on `develop`:

https://user-images.githubusercontent.com/273837/165919117-399765fc-99d8-4d31-99bc-33c179c049fd.mov

farmaazon committed:
QA fixes round 2

farmaazon commented:
Ad 4. Moved
Ad 7. Fixed for sure
Ad 10. This is by design; according to the design docs:
> Clicking on or pressing an Suggestion Acceptance Key while focusing a Component Module Group name inserts the name followed by a dot character into the Expression Input Panel, adds the name to the Component List Breadcrumb, and switches the Component List view to the items defined by the module.

Suggestion Acceptance Key is `tab`.
Ad 11. Fixed

akavel commented:
@farmaazon New issue found:

### 12

 1. Open the `component_group` debug scene.
 2. Press `up` key on the keyboard multiple times to move focus to the group name.
 3. Press `down` key on the keyboard once.
 4. Press `up` key on the keyboard once.

**Actual behavior:** the top entry is highlighted. After pressing `down` key again, still the top entry is highlighted.

**Expected behavior:** the group name is highlighted. After pressing `down` key again, the top entry is highlighted.

https://user-images.githubusercontent.com/273837/165957408-af065411-8c83-47d0-ad26-600f830befbc.mov

farmaazon committed:
QA fixes round 3

akavel reviewed (approved):
Please disable support for `PgUp`, `PgDn`, `Home`, and `End` keys on the keyboard. There are some corner cases around their behavior, but as we discussed, if we will support those keys, it will be done in the context of the full Component Group List, not the single Component Group view. Also, the Design Doc does not mention expected behaviors for those keys.

Other than that, LGTM from the Acc/QA Review point of view.

farmaazon committed:
Add test

mergify[bot] committed:
Merge branch 'develop' into wip/farmaazon/cgv-selectable-header-181725016

farmaazon committed:
Bump limit

farmaazon committed:
Prettier applied
